### PR TITLE
feat: Use truth vertex EDM in Examples

### PR DESCRIPTION
**Pull request by andiwand**
Currently truth vertices are not well defined. This is improved here by modifying the existing but unused truth vertex EDM and wiring it to the event generators and relevant performance writers.

blocked by
- https://github.com/acts-project/acts/pull/2904

### CODE REVIEW

File: Examples/Io/Root/src/RootParticleReader.cpp [hunk 1]
```diff
@@ -113,67 +112,55 @@ ActsExamples::RootParticleReader::~RootParticleReader() {
...
+    p.setProcess(static_cast<ActsFatras::ProcessType>((*m_process)[i]));
+    p.setPdg(static_cast<Acts::PdgParticle>((*m_particleType)[i]));
+    p.setCharge((*m_q)[i] * Acts::UnitConstants::e);
+    p.setMass((*m_m)[i] * Acts::UnitConstants::GeV);
+    p.setParticleId((*m_particleId)[i]);
+    p.setPosition4((*m_vx)[i] * Acts::UnitConstants::mm,
+                   (*m_vy)[i] * Acts::UnitConstants::mm,
+                   (*m_vz)[i] * Acts::UnitConstants::mm,
+                   (*m_vt)[i] * Acts::UnitConstants::mm);
+    // NOTE: depends on the normalization done in setDirection
```
Comment (pbutti) [inline Examples/Io/Root/src/RootParticleReader.cpp:152]: I'm not sure what this comment mean here in this context. Could you annotate for documentation?

File: Examples/Algorithms/GeneratorsPythia8/ActsExamples/Generators/Pythia8ProcessGenerator.cpp [hunk 2]
```diff
@@ -103,25 +108,30 @@ ActsExamples::SimParticleContainer ActsExamples::Pythia8Generator::operator()(
...
     // only secondaries have a defined vertex position
     if (m_cfg.labelSecondaries && genParticle.hasVertex()) {
       // either add to existing secondary vertex if exists or create new one
       // TODO can we do this w/o the manual search and position check?
       auto it = std::find_if(
-          vertexPositions.begin(), vertexPositions.end(),
-          [=](const SimParticle::Vector4& pos) { return (pos == pos4); });
-      if (it == vertexPositions.end()) {
+          vertices.begin(), vertices.end(),
+          [=](const SimVertex& other) { return pos4 == other.position4; });
```
[the remark this answers is not in the document]
Comment (andiwand) [reply Examples/Algorithms/GeneratorsPythia8/ActsExamples/Generators/Pythia8ProcessGenerator.cpp:121]: this is intentional so the behavior does not changes compared to the previous version of the code. In https://github.com/acts-project/acts/pull/3048 I relax that criteria to allow labeling secondary vertices "more correctly"

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Particle.hpp [hunk 3]
```diff
@@ -18,6 +18,11 @@
 namespace ActsExamples {
 namespace HepMC3Particle {
 
+/// @brief Returns the barcode translated into Acts
+/// @param particle HepMC3::GenParticle particle
+/// @return corresponding Acts barcode
```
[the remark this answers is not in the document]
Comment (andiwand) [reply Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Particle.hpp:23]: The HepMC stuff is not working properly. I touched it as little as possible to make things work. I don't think it makes sense to improve the situation in this PR

File: Examples/Io/Root/src/RootParticleReader.cpp (hunk 1 above)
Comment (andiwand) [reply Examples/Io/Root/src/RootParticleReader.cpp]: that comment was there before but I will try to improve it

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Particle.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (andiwand) [reply Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Particle.hpp:23]: https://github.com/acts-project/acts/issues/3050

**Review by pbutti — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [3eb0b4febd16ad8cee5664302e3542044ab17d35]
### :red_circle: Some tests have failed!
Please investigate the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/7104140)!

|status|job|report|
|------|---|------|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372223)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372222)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372221)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372220)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372219)||
|:green_circle: | [test_ActsConversionWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372218)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372217)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372216)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372215)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372214)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372213)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372212)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372211)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372210)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372209)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372208)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372207)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372206)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372205)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372204)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372203)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372202)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372201)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372200)||
|:red_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372199)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372198)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372197)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37372196)||

**Pull request merged**